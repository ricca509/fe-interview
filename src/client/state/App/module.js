export const selectBills = state => state.bills
  .filter(({ isBill }) => isBill)

export const selectPotentialBills = state => state.bills
  .filter(({ isBill }) => !isBill)

const fetchJson = url => {
  return fetch(url)
    .then(response => {
      console.log(response)

      if(response.ok) {
        return response;
      }
      throw new Error('Network response was not ok.');
    })
    .then(response => response.json())
    .catch(e => console.log(`Error fetching ${url}`, e))
};

export const fetchBills = () => {
  return fetchJson('http://localhost:3000/bills')
};