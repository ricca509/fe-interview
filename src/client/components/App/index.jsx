import React from 'react';
import Tabs from '../Tabs';
import Tab from '../Tab';
import Bills from '../Bills';
import { selectPotentialBills, selectBills, fetchBills } from '../../state/App/module';

const onBillSelected = (id) => {
  console.log('onBillSelected', id);
};

const onPotentialBillSelected = (id) => {
  console.log('onPotentialBillSelected', id);
};

class App extends React.Component {
  constructor() {
    super();

    this.state = {
      bills: [],
      loading: false,
      error: undefined
    };
  }

  componentDidMount()  {
    this.setState(() => ({
      loading: true
    }));

    fetchBills()
      .then(bills => {
        this.setState(() => ({
          bills,
          loading: false,
        }));
      })
      .catch(e => {
        console.log('Error');
        this.setState(() => ({
          loading: false,
          error: true
        }));
      })
  }

  render () {
    const { loading, error, bills } = this.state;
    if (loading) {
      return (
        <div>Loading...</div>
      );
    }

    if (error) {
      return (
        <div>Error!</div>
      );
    }

    if (!loading && !error && bills.length) {
      return (
        <Tabs>
          <Tab title="Bills">
            <Bills bills={selectBills(this.state)} onActionClicked={onBillSelected} />
          </Tab>
          <Tab title="Potential bills">
            <Bills bills={selectPotentialBills(this.state)} onActionClicked={onPotentialBillSelected} />
          </Tab>
        </Tabs>
      );
    }

    return false;
  }
}

export default App;