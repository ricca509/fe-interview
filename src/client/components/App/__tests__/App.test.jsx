import React from 'react';
import App from '../';
import { shallow } from 'enzyme';
import nock from 'nock';
import { fetchBills } from '../../../state/App/module';

jest.mock('../../../state/App/module')

describe('App component', ()=> {
  beforeEach(() => {
    fetchBills.mockImplementation(() => Promise.resolve({ bills: 'abc' }));
  });

  describe('when loading', ()=> {
    it('should render loading component', () => {
      const wrapper = shallow(<App />);
      wrapper.setState({
        bills: [],
        loading: true,
        error: undefined
      });

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when in error', ()=> {
    it('should render error component', () => {
      const wrapper = shallow(<App />);
      wrapper.setState({
        bills: [],
        loading: false,
        error: 'Not found'
      });

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when succesfully fetched the bills', ()=> {
    it('should render bills', () => {
      const wrapper = shallow(<App />);
      wrapper.setState({
        bills: [
          {
            "name": "Vodafone",
            "isBill": false,
            "transactions": [
              {
                "amount": 12.34,
                "date": "2017-01-13"
              }
            ],
            "id": "5a5caa1efe33900100fd8ed5"
          },
        ],
        loading: false,
        error: undefined
      });

      expect(wrapper).toMatchSnapshot();
    });
  });
});