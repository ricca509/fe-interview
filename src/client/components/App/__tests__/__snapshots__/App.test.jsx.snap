// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component when in error should render error component 1`] = `
<div>
  Error!
</div>
`;

exports[`App component when loading should render loading component 1`] = `
<div>
  Loading...
</div>
`;

exports[`App component when succesfully fetched the bills should render bills 1`] = `
<Tabs>
  <Tab
    title="Bills"
  >
    <Bills
      onActionClicked={[Function]}
    />
  </Tab>
  <Tab
    title="Potential bills"
  >
    <Bills
      onActionClicked={[Function]}
    />
  </Tab>
</Tabs>
`;
